test/e2e/vsphere_volume_placement.go

```diff
@@ -19,218 +19,412 @@ package e2e
 		c                  clientset.Interface
 		ns                 string
-		vsp                *vsphere.VSphere
-		volumePath         string
 		node1Name          string
-		node1LabelValue    string
 		node1KeyValueLabel map[string]string
-
 		node2Name          string
-		node2LabelValue    string
 		node2KeyValueLabel map[string]string
-
-		isNodeLabeled bool
+		testSeup           bool
 	)
-
-	/*
-		Steps
-		1. Create VMDK volume
-		2. Find two nodes with the status available and ready for scheduling.
-		3. Add labels to the both nodes. - (vsphere_e2e_label: Random UUID)
-
-	*/
-
 	BeforeEach(func() {
 		framework.SkipUnlessProviderIs("vsphere")
 		c = f.ClientSet
 		ns = f.Namespace.Name
 		framework.ExpectNoError(framework.WaitForAllNodesSchedulable(c, framework.TestContext.NodeSchedulableTimeout))
-
-		By("creating vmdk")
-		vsp, err := vsphere.GetVSphere()
-		Expect(err).NotTo(HaveOccurred())
-
-		volumePath, err = createVSphereVolume(vsp, nil)
-		Expect(err).NotTo(HaveOccurred())
-
-		if !isNodeLabeled {
-			nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
-			if len(nodeList.Items) != 0 {
-				node1Name = nodeList.Items[0].Name
-				node2Name = nodeList.Items[1].Name
-			} else {
-				framework.Failf("Unable to find ready and schedulable Node")
-			}
-			node1LabelValue = "vsphere_e2e_" + string(uuid.NewUUID())
-			node1KeyValueLabel = make(map[string]string)
-			node1KeyValueLabel["vsphere_e2e_label"] = node1LabelValue
-			framework.AddOrUpdateLabelOnNode(c, node1Name, "vsphere_e2e_label", node1LabelValue)
-
-			node2LabelValue = "vsphere_e2e_" + string(uuid.NewUUID())
-			node2KeyValueLabel = make(map[string]string)
-			node2KeyValueLabel["vsphere_e2e_label"] = node2LabelValue
-			framework.AddOrUpdateLabelOnNode(c, node2Name, "vsphere_e2e_label", node2LabelValue)
-
-		}
-
+		nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+		node1Name, node1KeyValueLabel, node2Name, node2KeyValueLabel = testSetupVolumePlacement(c, ns, nodeList)
+		testSeup = true
 	})
-
 	/*
 		Steps
 		1. Remove labels assigned to node 1 and node 2
 		2. Delete VMDK volume
 	*/
-
 	AddCleanupAction(func() {
-		if len(node1LabelValue) > 0 {
+		if len(node1KeyValueLabel) > 0 {
 			framework.RemoveLabelOffNode(c, node1Name, "vsphere_e2e_label")
 		}
-		if len(node2LabelValue) > 0 {
+		if len(node2KeyValueLabel) > 0 {
 			framework.RemoveLabelOffNode(c, node2Name, "vsphere_e2e_label")
 		}
-		if len(volumePath) > 0 {
-			vsp, err := vsphere.GetVSphere()
-			Expect(err).NotTo(HaveOccurred())
-			vsp.DeleteVolume(volumePath)
-		}
 	})
 
 	framework.KubeDescribe("provision pod on node with matching labels", func() {
+		var (
+			vsp         *vsphere.VSphere
+			volumePaths []string
+			err         error
+		)
+		/*
+			Steps
+			1. Create VMDK volume
+			2. Find two nodes with the status available and ready for scheduling.
+			3. Add labels to the both nodes. - (vsphere_e2e_label: Random UUID)
+
+		*/
+		BeforeEach(func() {
+			if !testSeup {
```

Comment:
What is the reason for this check? The top-level `KubeDescribe` 's `BeforeEach` does its setup and sets this bool. It will always be true here won't it?